src/trans/cpu/internal/dist_grid_ctl_mod.F90

```diff
@@ -94,178 +94,220 @@ SUBROUTINE DIST_GRID_CTL(PGPG,KFDISTG,KPROMA,KFROM,PGP,KSORT)
-  LLSAME=.TRUE.
-  IFROM=KFROM(1)
-  DO JFLD=2,KFDISTG
-    IF(KFROM(JFLD) /= IFROM) THEN
-      LLSAME=.FALSE.
-      EXIT
-    ENDIF
-  ENDDO
+ALLOCATE(ZRCV(D%NGPTOTMX,KFDISTG))
 
-  IMYFIELDS = 0
-  DO JFLD=1,KFDISTG
-    IF(KFROM(JFLD) == MYPROC) THEN
-      IMYFIELDS = IMYFIELDS+1
+IR=0
+! Post receives
+IOFFPROC(1)=0
+DO JROC=1,NPROC
+  IF (JROC /= MYPROC) THEN
+    IF (ANY(KFROM(1:KFDISTG) == JROC)) THEN
+      ITAG = MTAGDISTGP
+      IR=IR+1
+      CALL GSTATS(811,0)
+      CALL MPL_RECV(ZRCV(:,IOFFPROC(IR)+1:IOFFPROC(IR)+COUNT(KFROM(1:KFDISTG) == JROC)),&
+       & KMP_TYPE=JP_NON_BLOCKING_STANDARD,KREQUEST=IRCVREQ(IR), &
+       & KSOURCE=NPRCIDS(JROC),KTAG=ITAG,CDSTRING='DIST_GRID_CTL:')
+      CALL GSTATS(811,1)
+      IPROC(IR)=JROC
+      IOFFPROC(IR+1)=IOFFPROC(IR)+COUNT(KFROM(1:KFDISTG) == JROC)
     ENDIF
-  ENDDO
+  ENDIF
+ENDDO
 
-  CALL GSTATS(1663,0)
-  IF(IMYFIELDS > 0) THEN
-    ALLOCATE(ZBUF(D%NGPTOTMX,IMYFIELDS,NPROC))
-
-!$OMP PARALLEL DO SCHEDULE(DYNAMIC,1)&
-!$OMP&PRIVATE(JFLD,JA,JB,ISND,IGLOFF,IGL1,IGL2,IOFF,ILAST,J,&
-!$OMP&ILOFF,JGL,JLON)
-    DO JFLD=1,IMYFIELDS
-      DO JA=1,N_REGIONS_NS
-        DO JB=1,N_REGIONS(JA)
-          CALL SET2PE(ISND,JA,JB,0,0)
-
-          IGLOFF = D%NPTRFRSTLAT(JA)
-          IGL1   = D%NFRSTLAT(JA)
-          IGL2   = D%NLSTLAT(JA)
+! Fill and send
+ALLOCATE (ZBUF(D%NGPTOTMX,COUNT(KFROM(:) == MYPROC),NPROC))
+IF (SIZE(ZBUF) > 0) THEN
+  ! Force allocation right here, not inside the open-mp region below :
+  ZBUF(1,1,1)=HUGE(1._JPRB)
+ENDIF
+IS=0
+IFLDCOUNT=0
+DO JFLD=1,KFDISTG
+  IF(KFROM(JFLD) == MYPROC) THEN
+    IFLDCOUNT = IFLDCOUNT+1
+    IFLDLOC(IFLDCOUNT)=JFLD
+  ENDIF
+ENDDO
+IF(IFLDCOUNT > 0) THEN
+  DO JA=1,N_REGIONS_NS
+    IGLOFF = D%NPTRFRSTLAT(JA)
+    IGL1   = D%NFRSTLAT(JA)
+    IGL2   = D%NLSTLAT(JA)
+    DO JB=1,N_REGIONS(JA)
+      CALL SET2PE(ITO,JA,JB,0,0)
+      IF (ITO /= MYPROC) THEN
+        IF(JA == 1) THEN
           IOFF = 0
-          IF(JA > 1) THEN
-            IF( D%NLSTLAT(JA-1) == D%NFRSTLAT(JA) )THEN
-              ILAST = D%NLSTLAT(JA-1)-1
-            ELSE
-              ILAST = D%NLSTLAT(JA-1)
-            ENDIF
-            DO J=D%NFRSTLAT(1),ILAST
-              IOFF = IOFF+G%NLOEN(J)
-            ENDDO
+        ELSE
+          IF( D%NLSTLAT(JA-1) == D%NFRSTLAT(JA) )THEN
+            IOFF = SUM(G%NLOEN(D%NFRSTLAT(1):D%NLSTLAT(JA-1)-1))
+          ELSE
+            IOFF = SUM(G%NLOEN(D%NFRSTLAT(1):D%NLSTLAT(JA-1)))
           ENDIF
-          
-          ILEN(ISND,JFLD) = 0
-          ILOFF = 0
+        ENDIF
+        IOFFGP(IGL1) = 0
+        IOFFA(IGL1)=IOFF+D%NSTA(IGLOFF,JB)
+        DO JGL=IGL1,IGL2-1
+          ILONA(JGL)=D%NONL(IGLOFF+JGL-IGL1,JB)
+          IOFFGP(JGL+1) = SUM(D%NONL(IGLOFF:IGLOFF+JGL-IGL1,JB))
+          IOFFA(JGL+1)=IOFF+SUM(G%NLOEN(IGL1:JGL))+D%NSTA(IGLOFF+JGL-IGL1+1,JB)
+        ENDDO
+        ILONA(IGL2)=D%NONL(IGLOFF+IGL2-IGL1,JB)
+        CALL GSTATS(1663,0)
+!$OMP PARALLEL PRIVATE(JFLD,JGL)
```

Comment:
```suggestion
        !$OMP PARALLEL PRIVATE(JFLD,JGL)
```
Indentation for code readibility. I realise the previous file didn't indent `!$OMP` directives either, but we're gradually trying to improve the code standards... I'll suggest other places where these can be indented below.